Respect joint axes' initial_position

stefanbuettner opened the pull request:
SDF version 1.6 introduced an initial_position for joint axes. This is useful for loading models in a collision free pose. For example, Universal Robot's UR5e is penetrating the ground with the default initial configuration of all axes set to 0.0.

Fixes #2694

stefanbuettner commented:
It's still a work in progress because it doesn't seem to work quite right.

I tried it on the `double_pendulum_with_base` model with an `initial_position` of 1.58 for the `upper_joint` and -0.78 for the `lower_joint` but I get the following result:

![gazebo_initial_position_problem_double_pendulum](https://user-images.githubusercontent.com/13848659/90807878-6cf97300-e31f-11ea-8a2a-f8c316c7fc03.png)

Similar results on a UR10
![gazebo_initial_position_problem_ur10](https://user-images.githubusercontent.com/13848659/90810484-5228fd80-e323-11ea-9265-75b023ad00a8.png)

and a Kuka Youbot
![gazebo_initial_position_problem_youbot](https://user-images.githubusercontent.com/13848659/90810489-5523ee00-e323-11ea-9f98-321222927812.png)

Here are the model files I used:
[initial_position_test.zip](https://github.com/osrf/gazebo/files/5104836/initial_position_test.zip)

The first joint having the `initial_position` attribute in the chain seems to work alright but subsequent joints get messed up.

Is there something to be aware of during the model initialization process?

scpeters commented:
we've actually been planning to deprecate that tag in SDFormat 1.8 (see [osrf/sdformat bitbucket PR 683](https://osrf-migration.github.io/sdformat-gh-pages/#!/osrf/sdformat/pull-requests/683/page/1), since we figured the `<state>` tag would be more suitable for specifying initial joint positions.

* http://sdformat.org/spec?ver=1.7&elem=state#model_joint

cc @azeey

stefanbuettner commented:
Ok, but since it will be only deprecated and should still be supported in my opinion.

Is the state approach supported? I tried it with no effect.

chapulina commented:
> Is the state approach supported? I tried it with no effect.

Yes, I just checked with these quick steps:

1. Open `gazebo`
1. Insert a double pendulum
1. Let it move to a different position from its initial one
1. Pause and notice the pendulum configuration
1. Save the world
1. Close Gazebo
1. Load the saved world, paused so you can see the initial pose
1. The pendulum should start from the saved configuration

You can inspect the saved file to see how the `<state>` tag is used.

stefanbuettner commented:
Ok, I understood it wrong from reading the documentation. The steps you provided work. But I still think the `initial_position` tag should be fixed as well.

stefanbuettner commented:
Another thing I noticed: When loading a robot model, moving it in the desired configuration and saving the world, the state only contains `<link>` tags with `<pose>`, `<velocity>`, `<acceleration>`, and `<wrench>`. No `<joint>` tag. Is there a way to specify initial joint configurations? Because that's what I tried initially because it would be more intuitive in some situations.

chapulina commented:
I haven't tried, but there seems to be a `joint` tag inside state:

http://sdformat.org/spec?ver=1.7&elem=state#model_joint

stefanbuettner commented:
That's what I've tried as well but didn't work. This is my world file. I would expect the pendulum to start in a different configuration. But changing the angle has no effect.
Started with `gazebo --verbose --pause <world-file>`.
```xml
<?xml version="1.0" ?>
<sdf version="1.7">
  <world name="default">
    <include>
      <uri>model://ground_plane</uri>
    </include>

    <include>
      <uri>model://sun</uri>
    </include>

    <model name="double_pendulum">
      <include>
        <uri>model://double_pendulum_with_base</uri>
      </include>
    </model>

    <state world_name="default">
      <model name="double_pendulum">
        <joint name="double_pendulum_with_base::upper_joint">
          <angle axis="0">-1.58</angle>
        </joint>
      </model>
    </state>

  </world>
</sdf>
```

stefanbuettner committed:
Test if model states load the joint states

Spoiler alert: They don't

stefanbuettner committed:
Test if joint state is loaded for a model

Because it's currently not loaded.

stefanbuettner committed:
Fix joint state loading

Before the model state did not load the joint states even though the
code was there but only commented out.

This is part of the effort of supporting the SDF's joint state feature.

stefanbuettner committed:
Apply the loaded joint state

Use the commented code and update it to the new API.
Also extend it for more than one joint axis.

stefanbuettner commented:
I reverted the commit trying to fix the `initial_position` because it did not, so far.

The latest commits add support for the joint angle state. The example I posted before is working for me now. I also added regression tests.

stefanbuettner committed:
Test the initial_position of a joint axis

It was introduced in SDF version 1.6 and is not supported currently.

stefanbuettner committed:
Respect joint axes' initial_position

SDF version 1.6 introduced an initial_position for joint axes.
This is useful for loading models in a collision free pose.
Universal Robot's UR5e is penetrating the ground with the default
initial configuration of all axes set to 0.0.

stefanbuettner commented:
Now the `initial_position` tag also has the desired effect. Tests include only revolute joints, though.

stefanbuettner commented:
@chapulina do you mind having another look or shall we close it?

rfn123 commented:
Hi, 
sorry for bumping this, but will this be merged anytime soon?
I got the same problem and would need this fix :)
@scpeters @chapulina

monokaie commented:
@chapulina @azeey just listened to the ignition community meeting and thought about this PR when talking about initial states of the models.
